File: src/main/java/gregtechlite/gtlitecore/core/advancement/trigger/AdvancementTriggerImpl.java
package gregtechlite.gtlitecore.core.advancement.trigger;

import com.google.common.collect.Maps;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonObject;
import gregtechlite.gtlitecore.GTLiteMod;
import gregtechlite.gtlitecore.api.advancement.AdvancementCriterion;
import gregtechlite.gtlitecore.api.advancement.AdvancementTrigger;
import gregtechlite.gtlitecore.core.advancement.listener.AdvancementListeners;
import net.minecraft.advancements.PlayerAdvancements;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.ResourceLocation;
import org.jetbrains.annotations.ApiStatus.Internal;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

@Internal
public class AdvancementTriggerImpl<T extends AdvancementCriterion> implements AdvancementTrigger<T>
{

    private final ResourceLocation id;
    private final T criterion;
    private final Map<PlayerAdvancements, AdvancementListeners<T>> listeners = Maps.newHashMap();

    public AdvancementTriggerImpl(String name, @NotNull T criterion)
    {
        this.id = GTLiteMod.id(name);
        this.criterion = criterion;
    }

    @Override
    public void addListener(@NotNull PlayerAdvancements advancementsIn,
                            @NotNull Listener<T> listener)
    {
        AdvancementListeners<T> advancementListener = listeners.get(advancementsIn);
        if (advancementListener == null)
        {
            advancementListener = new AdvancementListeners<>(advancementsIn);
            listeners.put(advancementsIn, advancementListener);
        }
        advancementListener.add(listener);
    }

    @Override
    public void removeListener(@NotNull PlayerAdvancements advancementsIn,
                               @NotNull Listener<T> listener)
    {
        AdvancementListeners<T> advancementListener = listeners.get(advancementsIn);
        if (advancementListener != null)
        {
            advancementListener.remove(listener);
            if (advancementListener.isEmpty())
            {
                listeners.remove(advancementsIn);
            }
        }
    }

    @Override
    public void removeAllListeners(@NotNull PlayerAdvancements advancementsIn)
    {
        listeners.remove(advancementsIn);
    }

    @NotNull
    @Override
    public T deserializeInstance(@NotNull JsonObject object,
                                 @NotNull JsonDeserializationContext context)
    {
        return this.criterion;
    }

    @Override
    public void trigger(EntityPlayerMP player)
    {
        AdvancementListeners<T> listener = listeners.get(player.getAdvancements());
        if (listener != null)
        {
            listener.trigger(player);
        }
    }

    @NotNull
    @Override
    public ResourceLocation getId()
    {
        return this.id;
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/core/advancement/trigger/AdvancementTriggerImpl.kt
package gregtechlite.gtlitecore.core.advancement.trigger

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonObject
import gregtechlite.gtlitecore.GTLiteMod
import gregtechlite.gtlitecore.api.advancement.AdvancementCriterion
import gregtechlite.gtlitecore.api.advancement.AdvancementTrigger
import gregtechlite.gtlitecore.api.collection.openHashMapOf
import gregtechlite.gtlitecore.core.advancement.listener.AdvancementListeners
import net.minecraft.advancements.ICriterionTrigger
import net.minecraft.advancements.PlayerAdvancements
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.ResourceLocation

class AdvancementTriggerImpl<T : AdvancementCriterion>(private val id: ResourceLocation,
                                                       private val criterion: T) : AdvancementTrigger<T>
{
    private val listeners: MutableMap<PlayerAdvancements, AdvancementListeners<T>> = openHashMapOf()

    constructor(name: String, criterion: T) : this(GTLiteMod.id(name), criterion)

    override fun addListener(advancementsIn: PlayerAdvancements, listener: ICriterionTrigger.Listener<T>)
    {
        var _listener = listeners[advancementsIn]
        if (_listener == null)
        {
            _listener = AdvancementListeners(advancementsIn)
            listeners[advancementsIn] = _listener
        }
        _listener.add(listener)
    }

    override fun removeListener(advancementsIn: PlayerAdvancements, listener: ICriterionTrigger.Listener<T>)
    {
        val _listener = listeners[advancementsIn]
        _listener?.let {
            _listener.remove(listener)
            if (_listener.isEmpty())
            {
                listeners.remove(advancementsIn)
            }
        }
    }

    override fun removeAllListeners(advancementsIn: PlayerAdvancements)
    {
        listeners.remove(advancementsIn)
    }

    override fun deserializeInstance(json: JsonObject, context: JsonDeserializationContext): T = criterion

    override fun trigger(player: EntityPlayerMP?)
    {
        val listener = listeners[player?.advancements]
        listener?.let { listener.trigger(player) }
    }

    override fun getId(): ResourceLocation = id
}